File: USACO/Silver/Python/USACO 2015 Dec Silver Breed Counting.py
import sys

sys.stdin = open('bcount.in', 'r')
sys.stdout = open('bcount.out', 'w')

input = sys.stdin.readline

n, q = map(int, input().split())

a = [int(input()) for _ in range(n)]

ph, pg, pj = [0] * (n+1), [0] * (n+1), [0] * (n+1)

for i in range(n):
    ph[i+1] = ph[i]
    pg[i+1] = pg[i]
    pj[i+1] = pj[i]
    if a[i] == 1:
        ph[i+1] += 1
    elif a[i] == 2:
        pg[i+1] += 1
    else:
        pj[i+1] += 1

for _ in range(q):
    l, r = map(int, input().split())
    l -= 1
    print(ph[r]-ph[l], pg[r]-pg[l], pj[r]-pj[l])

File: USACO/Silver/C++/USACO 2015 Dec Silver Breed Counting.cpp
#include <bits/stdc++.h>

using namespace std;

#define ll long long

signed main() {
	freopen("bcount.in", "r", stdin);
	freopen("bcount.out", "w", stdout);

	int n, q;
	cin >> n >> q;
	vector<int> h(n+1), g(n+1), j(n+1);
	for(int i=0; i<n; ++i){
		int x; cin >> x;
		h[i+1] = h[i] + (x == 1);
		g[i+1] = g[i] + (x == 2);
		j[i+1] = j[i] + (x == 3);
	}

	while(q--){
		int a, b;
		cin >> a >> b;
		--a;
		cout << h[b] - h[a] << ' ' << g[b]-g[a] << ' ' << j[b]-j[a] << '\n';
	}
}
